### PR TITLE
Gh 74 docker maven plugin

### PR DESCRIPTION
**Pull request by go159lab**
Build and push docker images using google jib plugin without Dockerfile

### CODE REVIEW

File: gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/integrationtests/GraphControllerIT.java [hunk 1]
```diff
@@ -207,7 +207,7 @@ public void createGraph_shouldReturn500BadRequestWhenStoreTypeIsInvalidType() th
                 .content(inputJson)).andReturn();
 
         assertEquals(500, result.getResponse().getStatus());
-        final String expected = "{\"title\":\"RuntimeException\",\"detail\":\"StoreType is Invalid must be defined Valid Store Types supported are: accumuloStore, federatedStore, mapStore, proxyStore\"}";
+        final String expected = "{\"title\":\"RuntimeException\",\"detail\":\"StoreType is Invalid must be defined Valid Store Types supported are: proxyStore, mapStore, accumuloStore, federatedStore\"}";
```
Comment (r32575) [inline gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/integrationtests/GraphControllerIT.java:210]: Is this caused by the store types being output in a different order for different users? May want something more permanent if so. Could set `expected` to be a regular expression that captures the names. You can then split, order and compare to an `expectedStoreTypeList`.

File: gaffer-as-a-service/gaas-rest/pom.xml [hunk 2]
```diff
@@ -295,5 +295,44 @@
...
         </profile>
+        <profile>
+            <id>docker-build</id>
+            <activation>
+                <activeByDefault>false</activeByDefault>
+            </activation>
+            <build>
+            <plugins>
+                <plugin>
+                    <groupId>com.google.cloud.tools</groupId>
```
[the remark this answers is not in the document]
Comment (go159lab) [reply gaffer-as-a-service/gaas-rest/pom.xml:314]: @r32575 please review again I made changes accordingly to your comments.
 I put the following timestamp format
<maven.build.timestamp.format>yyyy-MM-dd.HH-mm-ss</maven.build.timestamp.format>
I tried different timestamp formats but docker is not allowing : Invalid image reference:

If you want I can try to change timestamp format

File: gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/integrationtests/GraphControllerIT.java (hunk 1 above)
Comment (go159lab) [reply gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/integrationtests/GraphControllerIT.java]: @r32575 please review again I made the changes as you requested.

File: gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/integrationtests/GraphControllerIT.java [hunk 3]
```diff
@@ -207,10 +213,30 @@ public void createGraph_shouldReturn500BadRequestWhenStoreTypeIsInvalidType() th
...
-        assertEquals(expected, result.getResponse().getContentAsString());
+
+        List<String> expectedStoreTypeList = new ArrayList<>(Arrays.asList("mapStore", "accumuloStore", "proxyStore", "federatedStore"));
+        Collections.sort(expectedStoreTypeList);
+
+        String contentAsString = result.getResponse().getContentAsString();
+        List<String> actualStoreTypeList =  findStoreTypes(contentAsString);
+        Collections.sort(actualStoreTypeList);
+
+        assertEquals(expectedStoreTypeList, actualStoreTypeList);
```
Comment (r32575) [inline gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/integrationtests/GraphControllerIT.java:224]: I was thinking more this:
```
final String[] expectedStoreType = new String[] {"accumuloStore", "federatedStore", "mapStore", "proxyStore"};
final String re = "\\{\"title\":\"RuntimeException\",\"detail\":\"StoreType is Invalid must be defined Valid Store Types supported are: (\\w+(,\\s\\w+)*)\"}";

final String contentAsString = result.getResponse().getContentAsString();

final Matcher matcher = Pattern.compile(re).matcher(contentAsString);

assertTrue(matcher.matches(), String.format("Regex: [%s] does not match string: [%s]", re, contentAsString));

final String group = matcher.group(1);
final String[] split = group.split(",\\s");
Arrays.sort(split);
assertArrayEquals(expectedStoreType, split);
```

File: gaffer-as-a-service/gaas-rest/pom.xml [hunk 4]
```diff
@@ -295,5 +296,63 @@
                 </plugins>
             </build>
         </profile>
+        <profile>
+            <id>docker-build</id>
+            <activation>
+                <activeByDefault>false</activeByDefault>
+            </activation>
+            <properties>
+                <buildType>buildTar</buildType>
```
Comment (r32575) [inline gaffer-as-a-service/gaas-rest/pom.xml:305]: Didn't realise you could put properties here too. I'm guessing that you can alter them from the command line?

File: gaffer-as-a-service/gaas-rest/pom.xml [hunk 5]
```diff
@@ -295,5 +296,63 @@
                 </plugins>
             </build>
         </profile>
+        <profile>
+            <id>docker-build</id>
```
Comment (r32575) [inline gaffer-as-a-service/gaas-rest/pom.xml:300]: Is the objective just to output the docker file? If that's the case then you could remove the profile and have it as a build plugin instead.

File: gaffer-as-a-service/gaas-rest/pom.xml (hunk 4 above)
Comment (go159lab) [reply gaffer-as-a-service/gaas-rest/pom.xml:305]: Sure. Command line properties value overwrites.

**Comment by go159lab** [comment]
@r32575 please review again. I made the changes as per your feedback.

File: gaffer-as-a-service/gaas-rest/README.md [hunk 6]
```diff
@@ -4,6 +4,18 @@ GaaS REST
 The Gaffer-as-a-Service REST API services requests to create, get and delete graphs.
 
 
+## Using jib plugin to build and push images
+The docker image built by Jib plugin you can change the default image using configuration
+
+#To build docker image locally
+
+mvn clean install jib:dockerBuild  -DbuildType=buildTar -DimageName= -DuserName= -DcredHelper=
```
[the remark this answers is not in the document]
Comment (go159lab) [reply gaffer-as-a-service/gaas-rest/README.md]: Done.

File: gaffer-as-a-service/gaas-rest/README.md [hunk 7]
```diff
@@ -4,6 +4,18 @@ GaaS REST
...
+
+
+# To build docker image and push to docker repository
+mvn clean install jib:build -DbuildType=buildTar -DimageName= -DuserName= -DcredHelper=
```
[the remark this answers is not in the document]
Comment (go159lab) [reply gaffer-as-a-service/gaas-rest/README.md]: Done. DbuildType=build  builds and push to dockerHub. Thank you

**Review by r32575 — APPROVED** [review]

**Pull request merged**